[READY] Launcher update + cumulative bug fix

ldeluigi opened the pull request:
### Changes
- Removed torrent related code
- Added launcher setting to run the server in test mode
- Refactored config to be a Singleton
- Updated default memory levels (1,000 MB client, 500 MB server)
- Added button to open up changelog
- Updated Beta URL
- Added UTF-8 encoding as default
- Force client and servers to be closed before performing an update
- Allow for system java to be used

Closes #1
Closes #6
Closes magefree/mage#5645
Closes magefree/mage#6462
Closes magefree/mage#7476

ldeluigi commented:
I added a replacement for backgorund number 6 because its absence caused random bugs.

See https://github.com/magefree/Launcher/blob/4a89e51a8502d99c7041a6ee83b2358967438bc9/src/main/java/com/xmage/launcher/XMageLauncher.java#L118

JayDi85 commented:
Can you remove all torrent related code too?

ldeluigi commented:
Done

JayDi85 commented:
Looks good. I'll test it later in different systems and situations.

P.S.
There is a [special test mode for server](https://github.com/magefree/mage/wiki/Development-Testing-Tools#start-server-by-script). And it require `-testMode=true` in app's command line, not java command line. So it can be useful to add checkbox near server settings to enable-disable test mode. With tooltip text like "Test mode allows you to quickly create a game with AI and customize any game situations and combos (use the cheat button on the player panel)"

JayDi85 commented:
Also no needs in -xm**s** settings, only -xm**x**

ldeluigi commented:
Done, plus solved another open issue.

ldeluigi reviewed (commented):
Related to what I've read in https://github.com/magefree/mage/issues/7515

ldeluigi commented:
I forgot to mention that I also solved https://github.com/magefree/mage/issues/4153

JayDi85 commented:
> Force the user to close clients before updating. 

Most "broken update" problems from users come from local server running.

JayDi85 commented:
> Change default option 'showClientConsole' to false

Nope, that's wrong. Its very important for new users to show logs on first run to troubleshooting the problems and to see startup progress (xmage can takes few minutes on first run on low end hdd computers). So logs must be visible by default.

ldeluigi commented:
@JayDi85 then please close this issue: https://github.com/magefree/mage/issues/4178

ldeluigi commented:
> > Force the user to close clients before updating.
> 
> Most "broken update" problems from users come from local server running.

Will force server close too then

ldeluigi commented:
@JayDi85 I think that we are good now, tell me if there is something more that's needed, or else I'm going to turn the pull request to "READY TO MERGE".


I just want to remark that someone should update (entirely) the Maven configuration because I don't know how Maven configuration works (and I don't have time to study it : P)

allentiak commented:
Ping?

Alex-Vasile commented:
What else needs to be done for this?

ldeluigi commented:
Probably the maven configuration
Feel free to fork my fork

Alex-Vasile commented:
Change line 8 from `<version>${version}</version>` to `<version>0.4.0</version>`.

For some reason IntelliJ maven doesn't tell you what's wrong, but if I run mvn from the command line it prints out the useful error:
`[ERROR]     Resolving expression: '${version}': Detected the following recursive expression cycle in 'version': [version] @ com.xmage:Launcher:${version},`

Alex-Vasile commented:
Also, can we take this opportunity to bump the java version for the launcher from 8 to at least 11 (if not 17)?

Alex-Vasile commented:
slf4j-log4j12 and maven-artifact dependancies could also use with a bump. IntelliJ will highlight them and show you new version. I don't know why dependabot isn't picking them up.

Alex-Vasile commented:
@ldeluigi, I've sent the pull request with the changes

ldeluigi commented:
Is there anything left to do?
@JayDi85

JayDi85 commented:
Yes, build/repackage script and java packages for distribution on multiple platforms. It’s my part.

ldeluigi commented:
Ok, let us know if you need any help : )
Thank you for your work

Hopefully the release of the new launcher goes in production in 2022 XD
It's been a year!

allentiak commented:
Hi, @JayDi85
Any news regarding this?

ldeluigi commented:
@Alex-Vasile I saw you committed to the main branch. I'm worried it could cause merge conflicts with this PR

Alex-Vasile commented:
> @Alex-Vasile I saw you committed to the main branch. I'm worried it could cause merge conflicts with this PR

I then commuted a merge to this PR to get rid of the conflict.

ldeluigi commented:
> > @Alex-Vasile I saw you committed to the main branch. I'm worried it could cause merge conflicts with this PR
> 
> I then did a merge to this PR to get rid of the conflict.

why then isn't this PR merged on github?

Alex-Vasile commented:
> > > @Alex-Vasile I saw you committed to the main branch. I'm worried it could cause merge conflicts with this PR
> > 
> > 
> > I then did a merge to this PR to get rid of the conflict.
> 
> why then isn't this PR merged on github?

I did a merge from magefree:master into ldeluigi:master (the branch for this PR).

There's some other things I want to wrap up on the XMage side (not the launcher) so that I can test the "launch with a user specified java" part of this PR. But I plan on merging this in the next few days so we can move on.

allentiak commented:
@Alex-Vasile
> I plan on merging this in the next few days so we can move on.

That's great to hear!

Really looking forward to that!

Alex-Vasile commented on src/main/java/com/xmage/launcher/XMageLauncher.java:
```diff
@@ -540,10 +504,31 @@ private boolean getConfig() {
         return false;
     }
 
+    private boolean checkJavaFX() {
+        try {
+            Class.forName("javafx.application.Platform");
+            return true;
+        } catch( ClassNotFoundException e ) {
+            return false;
+        }
+    }
+
```
I think this functionality is not needed (at least for the beta) as the Javafx dependancy has been removed. If it gets re-introduced in the future, it will be added to the client's pom.xml so that the project deals with its dependancies rather than hoisting that onto the user.

ldeluigi replied:
should I remove it entirely? or should I just comment line 509 for future reintroduction?

Alex-Vasile replied:
Leave it in there for now. I am not sure if we will need it or not, I have not yet looked into how to deal with javafx in the future if we switch to it over swing.

- [ ] comment out line 509
- [ ] Write a doctoring for `checkJava()`
- [ ] comment out calls to `checkJava()`

The last two will make it clearer what's going for anyone who comes by to look at it, and to work with it in the future.

ldeluigi replied:
As third option, I think you mean to comment out calls to `checkJavaFx()` inside `checkJava()`, right?

Alex-Vasile replied:
I originally meant checkJava, but it does seem to be doing something even if its set to not use system Java. I'll change the bullet point

ldeluigi replied:
If JavaFX is not needed I'd go for option 1 + option 3 which is always useful

Alex-Vasile replied:
2 and 3 are part of the same thing. For 2 I’m asking for a docstring overall what the function is doing, and for 3 I’m asking for the same information spread throughout the function so that anyone who needs to edit the function knows what each block of code is doing without having to read through it. I think both are needed (though 2 is just a summary of the points of 3).

ldeluigi replied:
Done

Alex-Vasile commented:
> Also, can we take this opportunity to bump the java version for the launcher from 8 to at least 11 (if not 17)?

In retrospect that was a mistake. I changed it back to 1.8 to not have this PR change the minimum requirement to run. This can be bumped at a later time if needed.

ldeluigi commented:
@Alex-Vasile I've refactored some code with IntelliJ automatic tools.
I've commented checkJava and commented out checkJavaFx implementation.
I also added a reminder message to user that change the XMage branch home to update XMage in order to apply changes.

Alex-Vasile commented:
It looks good, though work still needs to be done (separate PR) to get this to build (just the launcher) properly under Java 9+ (openjdk 11 & 18) won't let me build properly. They work fine for running, but don't work for building.

Alex-Vasile commented on src/main/java/com/xmage/launcher/XMageLauncher.java:
```diff
@@ -400,28 +340,32 @@ private static void openWebpage(String uri) {
 
     private void handleServer() {
         if (serverProcess == null) {
+            checkJava();
```
Alright, found a bug that needs fixing before merging (currently writing up changelog for this)


Steps:
1. Boot, tell it to use system JRE
2. Download the client and server
3. Switch from using the system JRE under settings
4. Click run client/server. When prompted about Java, click "OK" instead of "USE SYSTEM JRE"

Problem:
- The bundled java does **not** get downloaded, and the server won't run. The user must click "Update" manually in order to get the java to download

ldeluigi replied:
I'll add a reminder for the user + a check for that

Alex-Vasile replied:
Rather than a reminder, can you have it auto-download When the OK button is checked. (Or When the clients/server) try to be run?

ldeluigi replied:
It's not a good idea: the user should be able to explicitly decide when downloads should happen.
I'll add an error message that says something like "You need to update XMage first"

Alex-Vasile replied:
Then you need to:
1. Change the text on the pop up since it currently says that a new download will occur.
2. Grey out the launch client/server buttons the way they are greyed out when they aren't available

ldeluigi replied:
Done

Alex-Vasile replied:
Better, but still not there.

If the user changes the java setting (system -> downloaded) or (downloaded -> system) the check for an available java should happen when they click "Done" to close the settings.

E.g. **I switch from system JRE to downloaded one in settings**
Expected behaviour: 
- "Launch" buttons are greyed out if I don't have the java installed

Current behaviour:
- Launch buttons remain enabled
- Clicking on them asks me, again, if I would like to use my system JRE or the downloaded one. (it should not ask again since I **just** changed)
- I am prompted to update
- Buttons then get greyed out.

**Further change:**
- [ ] The "Update" button should say "Install" if either XMage or the downloaded Java are not installed.

ldeluigi replied:
Done, but I didn't rename the button in order to avoid confusions when messages refer to the "Update" button and users see "Install" instead

Alex-Vasile replied:
Still not working properly:

Alex-Vasile replied:
> Done, but I didn't rename the button in order to avoid confusions when messages refer to the "Update" button and users see "Install" instead

That's fine for now

Alex-Vasile replied:
Still not working properly:
- [ ] Switching from system java to downloaded java still displays a notification saying that "this is the first time the client" has been run. That notification shouldn't show up for the user since they have just switched java version
- [ ] Switching immediately back to system java does not re-enable the launch buttons

ldeluigi replied:
> Still not working properly:
> 
> * [x]  Switching from system java to downloaded java still displays a notification saying that "this is the first time the client" has been run. That notification shouldn't show up for the user since they have just switched java version
> * [x]  Switching immediately back to system java does not re-enable the launch buttons

Try now

Alex-Vasile replied:
> > Still not working properly:
> > 
> > * [x]   Switching from system java to downloaded java still displays a notification saying that "this is the first time the client" has been run. That notification shouldn't show up for the user since they have just switched java version
> > * [x]   Switching immediately back to system java does not re-enable the launch buttons
> 
> Try now

That all looks good!

Alex-Vasile replied:
> It's not a good idea: the user should be able to explicitly decide when downloads should happen.
> I'll add an error message that says something like "You need to update XMage first"

I want to come back to this since it's not sitting well with me, especially given the first run experience now. 

When the user opens Xmage for the firs time, they are told that XMage must download itself in order to play. Which I believe is the correct way to do this. The only reason they booted the launcher would be to play.

If the user also chooses to use the downloaded java (which they are prompted to choose **before** being told that xmage will download itself) then the JRE will also be downloaded automatically along with xmage.

The same should be true if the user switches from the system java to the downloaded one. When they switch to the downloaded one they are provided with a prompt telling them that there is no java installed. and asks them what do to. Letting them switch back to the installed java incase they don't want to download something is fine. But:
- [ ] Hitting okay should begin downloading the java. The user clicking okay has agreed that they want to use the downloaded java, and know that it isn't installed. There is no reason to make them click update separately.

ldeluigi replied:
Ok I'll do it your way

ldeluigi replied:
Try now

allentiak commented:
#14 has been finally merged!

Awesome!

allentiak commented:
Thanks to both @ldeluigi and @Alex-Vasile for your hard work!

allentiak commented:
I think the "multiple authors" GitHub feature fits great here.

https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors

Alex-Vasile commented:
> BTW: I think the "multiple authors" GitHub feature fits perfectly here ;-)
> 
> 
> 
> https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors

Well, I also meant to squash and merge, but forgot that I hadn't changed the default on the Launcher project (and only on Mage).

ldeluigi commented:
I think that the commit history got a bit corrupted... there's a huge phantom commit at the end of it

Alex-Vasile commented:
Alright, all cleaned up

allentiak commented:
Awesome!
Looking forward to using the new Launcher soon!

allentiak commented:
@JayDi85 @Alex-Vasile Any news regarding the release of the new version?

allentiak commented:
@JayDi85 @Alex-Vasile Any news regarding the release of the new version?

Alex-Vasile commented:
> @JayDi85 @Alex-Vasile Any news regarding the release of the new version?

I have no idea. I'm not active in the project anymore, sorry.

allentiak commented:
@JayDi85 Any news regarding the release of the new launcher?
